### PR TITLE
Added massiv-test-utils

### PR DESCRIPTION
**Pull request by CSVdB**
This package stores GenUnchecked and GenValid instances,
so that they can be imported separately by anyone writing tests
for their code which depends on Massiv.

### CODE REVIEW

File: massiv/src/Data/Massiv/Array/Manifest/Storable.hs [hunk 1]
```diff
@@ -30,21 +32,30 @@ import           Data.Massiv.Array.Unsafe            (unsafeGenerateArray,
...
 
 -- | Representation for `Storable` elements
-data S = S deriving Show
+data S = S deriving (Show, Generic)
+
+instance Validity S
 
 type instance EltRepr S ix = M
 
+instance (MVS.Storable e, Validity e) => Validity (VS.Vector e) where
```
Comment (lehins) [inline massiv/src/Data/Massiv/Array/Manifest/Storable.hs]: I wonder why `validity-vector` doesn't have an instance for Storable (and other) vectors, it would definitely make more sense there.

File: massiv/src/Data/Massiv/Core/Index/Ix.hs [hunk 2]
```diff
@@ -93,6 +100,9 @@ type family Ix (n :: Nat) = r | r -> n where
 data IxN (n :: Nat) where
   (:>) :: Dimensions (Ix (n - 1)) ~ (n - 1) => {-# UNPACK #-} !Int -> !(Ix (n - 1)) -> IxN n
 
+instance Validity (Ix (n - 1)) => Validity (IxN n) where
+  validate (a :> b) = declare "first element is positive" (a > 0) <> delve "tail" b
```
Comment (lehins) [inline massiv/src/Data/Massiv/Core/Index/Ix.hs]: I don't think `Validity` instance for `IxN` should be different for different versions of GHC (I just noticed two instance declarations). Also, correct me if I am wrong, but `IxN` can have first element equal to `0` (eg. `0 :> 0 :. 1`) , since it is not only used for size but for indexing as well. Moreover there, there are places in `Stencil`, as you know, that have negative indices, but I guess as far as `Validity` is concerned it might not be important.

File: massiv/src/Data/Massiv/Core/Index/Class.hs [hunk 3]
```diff
@@ -19,14 +21,21 @@ module Data.Massiv.Core.Index.Class where
...
 import           Data.Massiv.Core.Iterator
+import           Data.Validity
 import           GHC.TypeLits
+import           GHC.Generics (Generic)
 
 -- | A way to select Array dimension.
 newtype Dim = Dim Int deriving (Show, Eq, Ord, Num, Real, Integral, Enum)
 
+instance Validity Dim where
+  validate (Dim n) = declare "A Dim is positive" $ n >= 0
```
Comment (lehins) [inline massiv/src/Data/Massiv/Core/Index/Class.hs]: `Dim` indeed is positive, since we don't deal with zero dimensional arrays, so shouldn't it be `n > 0`? At least, as far as I remember, there is no place in the package, where n = 0 is considered valid.

**Comment by lehins** [comment]
@CSVdB, to be honest with you, I was also thinking at first having a separate package for testing with QuickCheck instances, but I am really not fond of having a separate package that's useful only testing. Here is a [short example with a solution](https://gist.github.com/lehins/3498c5eb3bd5ca82af47c740c1c37256#file-motivation-md) I was thinking about for a while.
Please, let me know what you think of that.
cc @NorfairKing, I'd really love to hear you opinion on that as well.

**Comment by lehins** [comment]
@CSVdB Disregard my last comment. Having put some more thought into that approach using cabal flags can create more problems than it is worth.

Putting that aside, now I think it would make sense to move QuickCheck related data types and instances into  `massiv-test-utils` as well, so they can also be used by other packages, what are your thoughts on that?

**Comment by NorfairKing** [comment]
@lehins I'm happy you reconsidered.
Just to be clear, though, `Validity` is not just for testing, but  `GenUnchecked` and `GenValid` are.

I could have a look at the instances in this PR as well, if you'd like.

**Comment by lehins** [comment]
@NorfairKing I still don't like the fact that we need extra package just for instances that are needed purely for testing, it's unnecessary maintenance overhead, in my opinion. But, unfortunately, this the best solution we have so far, so I'll stick to it.

I certainly can see `Validity` and `RelativeValidty` be useful on their own outside of testing. For example in massiv, there could be an instance:

```haskell
instace Size r ix e => RelativeValidity ix (Array r ix e) where
  isValidFor ix arr = isSafeIndex (size arr) ix
```
which would allow to check if an index is legit for this array.

File: massiv-test-utils/src/Test/Massiv/Core/Computation.hs [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+module Test.Massiv.Core.Computation where
+
+import Import
+
+import Test.QuickCheck
+
+import Data.Massiv.Core
+
+instance GenUnchecked Comp where
+    genUnchecked = oneof [ParOn <$> genUnchecked, pure Seq]
```
Comment (NorfairKing) [inline massiv-test-utils/src/Test/Massiv/Core/Computation.hs]: This can be generated.

File: massiv-test-utils/src/Test/Massiv/Core/Computation.hs [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    shrinkUnchecked = const []
```
Comment (NorfairKing) [inline massiv-test-utils/src/Test/Massiv/Core/Computation.hs]: That shouldn't be necessary

File: massiv-test-utils/src/Test/Massiv/Core/Computation.hs [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+instance GenValid Comp where
+    genValid = oneof [ParOn <$> genValid, pure Seq]
```
Comment (NorfairKing) [inline massiv-test-utils/src/Test/Massiv/Core/Computation.hs]: This can be generated.

File: massiv-test-utils/src/Test/Massiv/Core/Index/Ix.hs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+import Data.Massiv.Core
+
+intToList :: Int -> [Int]
+intToList x =
+    if x >= 0
+        then [0 .. x - 1]
+        else reverse [x + 1 .. 0]
+
+instance GenUnchecked Ix0 where
```
Comment (NorfairKing) [inline massiv-test-utils/src/Test/Massiv/Core/Index/Ix.hs]: This can be generated

File: massiv-test-utils/src/Test/Massiv/Core/Index/Ix.hs [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    genUnchecked = pure Ix0
+
+instance GenValid Ix0 where
+    genValid = pure Ix0
```
Comment (NorfairKing) [inline massiv-test-utils/src/Test/Massiv/Core/Index/Ix.hs]: This can be generated

File: massiv-test-utils/src/Test/Massiv/Core/Index/Ix.hs [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+instance GenUnchecked Ix2 where
+    genUnchecked = (:.) <$> genUnchecked <*> genUnchecked
+    shrinkUnchecked (a :. b) = [i :. j | i <- intToList a, j <- intToList b]
```
Comment (NorfairKing) [inline massiv-test-utils/src/Test/Massiv/Core/Index/Ix.hs]: This can be generated

File: massiv-test-utils/src/Test/Massiv/Core/Index/Ix.hs [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+instance GenValid Ix2 where
+    genValid = (:.) <$> genValid <*> genValid
```
Comment (NorfairKing) [inline massiv-test-utils/src/Test/Massiv/Core/Index/Ix.hs]: This can be generated

File: massiv-test-utils/src/Test/Massiv/Core/Index/Ix.hs [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+        else reverse [x + 1 .. 0]
+
+instance GenUnchecked Ix0 where
+    genUnchecked = pure Ix0
+
+instance GenValid Ix0 where
+    genValid = pure Ix0
+
+instance GenUnchecked Ix2 where
+    genUnchecked = (:.) <$> genUnchecked <*> genUnchecked
```
Comment (NorfairKing) [inline massiv-test-utils/src/Test/Massiv/Core/Index/Ix.hs]: This can be generated

File: massiv-test-utils/src/Test/Massiv/Core/Index/Ix.hs (hunk 9 above)
Comment (NorfairKing) [inline massiv-test-utils/src/Test/Massiv/Core/Index/Ix.hs]: This is incorrect, you need to shrink the tuple together.

File: massiv-test-utils/src/Test/Massiv/Core/Index/Ix.hs [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    shrinkUnchecked (a :. b) = [i :. j | i <- intToList a, j <- intToList b]
+
+instance GenValid Ix2 where
+    genValid = (:.) <$> genValid <*> genValid
+
+instance GenUnchecked (Ix (n - 1)) => GenUnchecked (IxN n) where
+    genUnchecked = (:>) <$> genUnchecked <*> genUnchecked
```
Comment (NorfairKing) [inline massiv-test-utils/src/Test/Massiv/Core/Index/Ix.hs]: This can be generated

File: massiv/src/Data/Massiv/Array/Delayed/Windowed.hs [hunk 13]
```diff
@@ -23,16 +23,21 @@ import           Control.Monad                          (when)
...
 import           Data.Massiv.Core.Scheduler
 import           Data.Proxy                             (Proxy (..))
 import           Data.Typeable                          (showsTypeRep, typeRep)
+import           Data.Validity
 
 -- | Delayed Windowed Array representation.
 data DW = DW
 
+instance Validity DW where
+    validate = trivialValidation
```
Comment (NorfairKing) [inline massiv/src/Data/Massiv/Array/Delayed/Windowed.hs]: This can be generated.

File: massiv/src/Data/Massiv/Array/Manifest/BoxedNF.hs [hunk 14]
```diff
@@ -35,23 +39,36 @@ import           Data.Massiv.Core.Common
...
+
+instance Validity N
 
 type instance EltRepr N ix = M
 
+instance Validity a => Validity (A.Array a) where
+#if MIN_VERSION_primitive(0,6,2)
+    validate = foldMap validate
+#else
+    validate = trivialValidation
```
Comment (NorfairKing) [inline massiv/src/Data/Massiv/Array/Manifest/BoxedNF.hs]: Again: this is wrong, but no need to fix it if you can replace the dependency with the newest version of `validity-array` soon.

File: massiv/src/Data/Massiv/Array/Manifest/Internal.hs [hunk 15]
```diff
@@ -61,6 +63,17 @@ data M
...
+
+instance (Validity ix, Index ix, Validity e) => Validity (Array M ix e) where
+    validate MArray {..} = mconcat
+        [ delve "comp" mComp
+        , delve "size" mSize
+        , checkUnsafeLinearIndex (totalElem mSize) mUnsafeLinearIndex
+        ]
+
+checkUnsafeLinearIndex :: Validity e => Int -> (Int -> e) -> Validation
+checkUnsafeLinearIndex n func = foldMap (validate . func) $ [0..n - 1]
```
Comment (NorfairKing) [inline massiv/src/Data/Massiv/Array/Manifest/Internal.hs]: you still need to decorate this for a nice error message.

File: massiv/src/Data/Massiv/Array/Manifest/Storable.hs (hunk 1 above)
Comment (NorfairKing) [reply massiv/src/Data/Massiv/Array/Manifest/Storable.hs]: Coming soon.

File: massiv/src/Data/Massiv/Core/Computation.hs [hunk 16]
```diff
@@ -35,6 +36,10 @@ data Comp
   -- defaults to number available cores.
   deriving (Show, Eq)
 
+instance Validity Comp where
+    validate Seq = valid
+    validate (ParOn x) = delve "ParOn" x
```
Comment (NorfairKing) [inline massiv/src/Data/Massiv/Core/Computation.hs]: This can be generated.

File: massiv/src/Data/Massiv/Core/Index.hs [hunk 17]
```diff
@@ -77,6 +78,13 @@ data Border e =
               --
   deriving (Eq, Show)
 
+instance Validity e => Validity (Border e) where
+    validate (Fill e) = delve "Fill" e
+    validate Wrap = valid
+    validate Edge = valid
+    validate Reflect = valid
+    validate Continue = valid
```
Comment (NorfairKing) [inline massiv/src/Data/Massiv/Core/Index.hs]: This can be generated

File: massiv/src/Data/Massiv/Core/Index/Class.hs (hunk 3 above)
Comment (NorfairKing) [reply massiv/src/Data/Massiv/Core/Index/Class.hs]: @lehins Why not store a `Word` or `Natural` then? (optionally with the semantics that `Dim 0` means `1`?

File: massiv/src/Data/Massiv/Core/Index/Ix.hs [hunk 18]
```diff
@@ -50,6 +51,9 @@ pattern Ix1 i = i
 -- | 2-dimensional index. This also a base index for higher dimensions.
 data Ix2 = (:.) {-# UNPACK #-} !Int {-# UNPACK #-} !Int
 
+instance Validity Ix2 where
+  validate = trivialValidation
```
Comment (NorfairKing) [inline massiv/src/Data/Massiv/Core/Index/Ix.hs]: This can be generated.

File: massiv/src/Data/Massiv/Core/Index/Ix.hs [hunk 19]
```diff
@@ -81,6 +85,9 @@ pattern Ix5 i j k l m = i :> j :> k :> l :. m
 -- | n-dimensional index. Needs a base case, which is the `Ix2`.
 data IxN (n :: Nat) = (:>) {-# UNPACK #-} !Int !(Ix (n - 1))
 
+instance Validity (Ix (n - 1)) => Validity (IxN n) where
+  validate (_ :> b) = delve "tail" b
```
Comment (NorfairKing) [inline massiv/src/Data/Massiv/Core/Index/Ix.hs]: Is there a reason that the head doesn't need to be validated?

File: massiv/src/Data/Massiv/Core/Index/Ix.hs (hunk 2 above)
Comment (NorfairKing) [reply massiv/src/Data/Massiv/Core/Index/Ix.hs]: > Moreover there, there are places in Stencil, as you know, that have negative indices, but I guess as far as Validity is concerned it might not be important.

If `IxN` with a `0` or a negative value is _ever_ a valid thing, then the validity instance shouldn't try to validate against that.

File: massiv/src/Data/Massiv/Core/List.hs [hunk 20]
```diff
@@ -36,11 +37,16 @@ import           Data.Massiv.Core.Common
...
 import           Data.Typeable
+import           Data.Validity
 import           GHC.Exts
+import           GHC.Generics (Generic)
 import           System.IO.Unsafe           (unsafePerformIO)
 
 data LN
 
+instance Validity LN where
+    validate = trivialValidation
```
Comment (NorfairKing) [inline massiv/src/Data/Massiv/Core/List.hs]: This _should_ be generatable. It doesn't really matter because you can never call it anyway, but best not to make extra code if we can avoid it.

File: massiv/src/Data/Massiv/Core/Index/Class.hs (hunk 3 above)
Comment (lehins) [reply massiv/src/Data/Massiv/Core/Index/Class.hs]: Using `Word` doesn't protect you from negative values, it simply wraps around, which is terrible in my opinion. Attempt to construct a negative `Natural` results in a runtime underflow exception, which is not much better. In every place `Dim` is used, an invalid value will result in a sensible error message. Not only negative values will result in an error, but in case `Dim` being higher than dimensionality of the array will as well. Plus there are functions that will return `Nothing` in above cases, which would become impossible for `Natural` in pure setting.

**Comment by lehins** [comment]
Considering that we are creating a separate `massiv-test-utils` package, I would prefer all of the validity instances to be moved there as well, this way there is no dependency on `validity`, since it isn't necessary for proper operation on massiv.

File: massiv/src/Data/Massiv/Core/Index/Class.hs (hunk 3 above)
Comment (lehins) [reply massiv/src/Data/Massiv/Core/Index/Class.hs]: By the way, this reminds me that there could be instances as well: `Index ix => RelativeValidity Dim (Array r ix e)` and `Index ix => RelativeValidity Dim ix`

File: massiv/src/Data/Massiv/Array/Manifest/Storable.hs (hunk 1 above)
Comment (CSVdB) [reply massiv/src/Data/Massiv/Array/Manifest/Storable.hs]: I also made that PR to validity.
Now that it's launched, I'll remove the relevant validity instances and update this PR.

File: massiv-test-utils/src/Test/Massiv/Core/Index/Ix.hs (hunk 10 above)
Comment (CSVdB) [reply massiv-test-utils/src/Test/Massiv/Core/Index/Ix.hs]: It can, but the generated version is different from mine.
Generated: genValid = genUnchecked `shouldSatisfy` isValid
  where
    genUnchecked = (:.) <$> genUnchecked <*> genUnchecked
If "isValid" is not always true, my version will be faster.

File: massiv/src/Data/Massiv/Array/Delayed/Windowed.hs (hunk 13 above)
Comment (CSVdB) [reply massiv/src/Data/Massiv/Array/Delayed/Windowed.hs]: Yes, but then I would have to add the "DeriveGenerics" language extension, import GHC.Generics, and derive a Generic description for DW. This sounded like a bad idea.
Do you agree?

File: massiv/src/Data/Massiv/Core/Index/Ix.hs (hunk 19 above)
Comment (CSVdB) [reply massiv/src/Data/Massiv/Core/Index/Ix.hs]: The head is an Int, which is always valid.
I could make this "trivialValidation" though.
The reason I delved into the tail, is for maintainability:
If you put a restriction on Ix0, Ix1, Ix2 later on, this would cause unexpected behavior
in higher dimensional index types.

File: massiv/src/Data/Massiv/Core/List.hs (hunk 20 above)
Comment (CSVdB) [reply massiv/src/Data/Massiv/Core/List.hs]: You're not actually saving code by generating it. You need to add DeriveGeneric, GHC.Generics and "deriving Generic" (for LN) before you can derive it.

File: massiv-test-utils/src/Test/Massiv/Core/Index/Ix.hs (hunk 10 above)
Comment (NorfairKing) [reply massiv-test-utils/src/Test/Massiv/Core/Index/Ix.hs]: Ah right, fair point.

File: massiv/src/Data/Massiv/Array/Delayed/Windowed.hs (hunk 13 above)
Comment (NorfairKing) [reply massiv/src/Data/Massiv/Array/Delayed/Windowed.hs]: I don't agree, but it's not up to me.

File: massiv/src/Data/Massiv/Core/List.hs (hunk 20 above)
Comment (NorfairKing) [reply massiv/src/Data/Massiv/Core/List.hs]: Same comment as above, up to @lehins

File: massiv/src/Data/Massiv/Core/Index/Class.hs (hunk 3 above)
Comment (lehins) [reply massiv/src/Data/Massiv/Core/Index/Class.hs]: That's precisely the point. `Ix0` does NOT have an instance of `Index` class and is not considered a valid index by the library. It is only needed only for the type system to be happy, since `Lower Ix1 = Ix0`. For the same reason, there is no place in the library where `Dim 0` would be valid either.

File: massiv/src/Data/Massiv/Array/Delayed/Windowed.hs (hunk 13 above)
Comment (lehins) [reply massiv/src/Data/Massiv/Array/Delayed/Windowed.hs]: TLDR; @CSVdB, I'm ok with either approach. Same applies to all spots where @NorfairKing suggested generation.

I was thinking about adding Generic instance to most of the data types in massiv anyways, since it is pretty common for users of libraries to want them for various reasons. So I don't oppose to adding generic instances, but I don't find it important to switch to derived `Validity` instances for two reasons:
* you already have code that works and is semantically equivalent to the derived instances (at least I'd hope so ;)).
* Using derived mechanism does sometime marginally slow down compilation. Not terribly important.

**Comment by CSVdB** [comment]
I would argue the validity instances don't belong in the test utils package.
Of course if you disagree, I will move the instances.

The idea behind the test utils package is that it contains everything which people only need for testing purposes. Validity has reasons for existing outside the scope of just tests.
Apart from being non-ideal to keep the validity instances in the test utils package,
it would force anyone who wants to use these instances in their library (and not just in the tests)
to make their library depend on testing-specific libraries like QuickCheck and Hspec.
This situation would be far from ideal, in my opinion.

What do you think about this argument?

**Comment by lehins** [comment]
My biggest argument for validity instances to not be included in massiv directly, is that I don't see a use case for them outside testing. As such it becomes an unnecessary dependency.

> Validity has reasons for existing outside the scope of just tests.

I can see that being a case in some situations for other packages, but could you provide some examples of this in massiv? 

As far as:

> anyone who wants to use these instances in their library

If that will ever happen, we can always release `validity-massiv` and follow the common pattern already established for other libraries.

File: massiv/src/Data/Massiv/Array/Manifest/Internal.hs [hunk 21]
```diff
@@ -61,6 +63,17 @@ data M
 data instance Array M ix e = MArray { mComp :: !Comp
                                     , mSize :: !ix
                                     , mUnsafeLinearIndex :: Int -> e }
+
+instance (Validity ix, Index ix, Validity e) => Validity (Array M ix e) where
+    validate MArray {..} = mconcat
+        [ delve "comp" mComp
+        , delve "size" mSize
+        , decorate "Array M elements" $ checkUnsafeLinearIndex (totalElem mSize) mUnsafeLinearIndex
```
Comment (lehins) [inline massiv/src/Data/Massiv/Array/Manifest/Internal.hs]: You could use `validateFunc 0` here instead of `checkUnsafeLinearIndex`.

**Comment by CSVdB** [comment]
So in conclusion, where should the validity instances go?
Shall I put them into a new massiv-validity package immediately?

**Comment by lehins** [comment]
@CSVdB Yes, separate `massiv-validity` would be just fine. Or maybe `validity-massiv` would be even better ;) Just so the pattern with other packages is preserved.

**Comment by CSVdB** [comment]
I'm fixing the code right now, though there is one problem:
Validity requires the internals of some of the arrays, which are not exposed.
Possible solutions:
- Expose every module under a name which involves "internal" or "unsafe"
- Add the validity instances to massiv itself anyway

What do you think?

**Comment by lehins** [comment]
@CSVdB I new about that problem even before I proposed to move validity instance to a separate package and I thought about the solution as well.

This is way too drastic and very ugly:
> * Expose every module under a name which involves "internal" or "unsafe"

There is already an exposed module `Data.Massiv.Array.Unsafe` that re-exports almost all of the unsafe functionality. It is done as separate module to promote safety of the library.

That's what we are trying to solve here:
> * Add the validity instances to massiv itself anyway

There is also a third option:
* Export just the accessors to the internals of data types.

I'll do some work on that tomorrow and will let you know where we can go from there.

**Comment by lehins** [comment]
@CSVdB I've made some changes to massiv that include exported functions, which allow access to the insides of all array types, except `DW` and `Stencil`. Latter two are really special cases and I need to think a little what can be exported and what does even mean for those things to be valid. Actually, same concern applies to `D` and `N` array types, more on that below.

Changes I am talking about can be found in PR #39 , which will be merged into master shortly.

Here are a few examples that show how to access the insides:

```
instance (Validity ix, Index ix) => Validity (Array P ix e) where
    validate arr = delve "comp" (getComp arr) <> delve "size" (getSize arr) <> delve "data" (toByteArray arr)
```

```
instance Validity (Nested LN ix e) => Validity (Array L ix e)
    validate arr = delve "comp" (getComp arr) <> delve "data" (toNested arr)
instance Validity (Nested LN ix e) => Validity (Array LN ix e)
    validate arr = delve "comp" (getComp arr) <> delve "data" (toNested arr)
```

```
instance (Index ix, Validity ix, Validity e) => Validity (Array D ix e) where
    validate arr =
        mconcat
            [ delve "comp" (getComp arr)
            , delve "size" (getSize arr)
            , decorate "array" $ validateFunc (pureIndex 0) (getSize arr) (unsafeIndex arr)
            ]
```

The last instance I am not quite sure about, since it validates a function, which is sort of impossible to validate. For example, how do we know that it won't read out of bounds memory:
```
λ> let arr = makeArrayR S Seq (Ix1 3) id
λ> arr
(Array S Seq (3)
  [ 0,1,2 ])
λ> let badArr = makeArrayR D Seq (Ix1 5) (unsafeIndex arr)
λ> badArr
(Array D Seq (5)
  [ 0,1,2,139895927297944,52 ])
```
The values it returns might be valid, but the array is certainly invalid.

**Comment by CSVdB** [comment]
I did some of the work you were asking for.
However, I do have one problem: I can't seem to find the `getSize` function you are referring to.
Am I missing something?

**Comment by lehins** [comment]
I apologize, I meant just `size` function.10:29, July 26, 2018, NickVdB <notifications@github.com>:I did some of the work you were asking for.
However, I do have one problem: I can't seem to find the getSize function you are referring to.
Am I missing something?

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

Отправлено из мобильной Яндекс.Почты: http://m.ya.ru/ymail

**Comment by lehins** [comment]
This PR looks dead. @CSVdB I am inclined to closing it, do you still have any plans for this?
